### PR TITLE
[accesscontrol] Add dependency dag

### PR DESCRIPTION
**Pull request by udpatil**

### CODE REVIEW

File: go.mod
Comment (BrandonWeng) [inline go.mod]: Reminder to revert this before merging

File: app/app.go
Comment (BrandonWeng) [inline app/app.go:379]: for my understanding, why was this commented out previously?

File: app/app.go
Comment (BrandonWeng) [inline app/app.go:972]: why not just check for `ErrCycleInDAG` or maybe we could have `if err != nil` or `if err != nil || err == ErrCycleInDAG`?  I think it'll be a bit easier to read

File: app/graph.go
Comment (BrandonWeng) [inline app/graph.go]: Why do we need to store the access operation of the edge here? If we have A -> B, will the access operation in the edge refer to the operation in A or B?

File: app/graph.go
Comment (BrandonWeng) [inline app/graph.go]: nit - could we get some high level comment on how this works

File: app/graph.go
Comment (BrandonWeng) [inline app/graph.go]: thoughts on making this a helper? Seems like the logic is repeated a couple times

File: app/graph.go
Comment (BrandonWeng) [inline app/graph.go]: I believe there's also `AccessType_ANY`, though I think that might be similar to Accesstype_WRITE since it should just block on previous reads and writes

File: app/graph.go
Comment (BrandonWeng) [reply app/graph.go]: Since we're adding the edge to the previous accessOperation (either read or write), would this be an issue when that transaction is not the last operation in a given message 

(i.e don't we need to wait for write operations to commit in a message before reading it?)

**Review by BrandonWeng — COMMENTED** [review]
thanks for the change, just have some questions

File: app/app.go
Comment (udpatil) [reply app/app.go:379]: I think this was introduced in a previous change Tony made and then commented out, I needed this field for the dependency DAG so we can decode TXs to be able to get their message information to get dependencies.

File: app/app.go
Comment (udpatil) [reply app/app.go:972]: if it is a different error, we might need to handle the tx processing differently (currently not present). I can leave a TODO so that if its a different error we can have special handling for that.

File: app/graph.go
Comment (udpatil) [reply app/graph.go]: the reasoning is because of cases where we have an edge for a write operation that occurs earlier in a Message, we want it to actually be linked to the very last access operation instead so that the completion signal can be sent when the msg processing is over. heres an example:
In this case, we have message 1 writing to resource A, but the actual edge would be from the read of resource B since thats the last access op of the message, so once that is complete we can know that the message is committed. 
m1wA -> m1rB ----> m2rA

 maybe it might be better instead of implicitly linking from the last access op of a message, each message can have an accessOp specifying the end of the message (This becomes more relevant if the last access op (eg m1rB in this case) is prior to end of message and there can still be a panic in between that and the end of message handling. This way, if we have a write in the message, we send its completion signal from the end of message access operation instead of the last specified access operation. WDYT?

File: app/graph.go
Comment (udpatil) [reply app/graph.go]: for this specifically, if we are dependent on a previous write, we actually are tracking the last operation in a message and linking edge from that instead of directly linked to the write access op. I'll go back and add some more comments to discuss in more detail though so its clearer.

File: app/graph.go
Comment (udpatil) [reply app/graph.go]: yeah, I already have a TODO above to replace with a helper that also takes into account resource hierarchy so instead of only linking resources of the same type, it'll also link from resources of superseding types as well.

File: app/graph.go
Comment (udpatil) [reply app/graph.go]: Good point, I'll modify it to handle this as well, I was initially assuming the `0` proto values were being treated as basically invalid but this is a more graceful way of handling it.

Comment (udpatil) [reply app/graph.go]: wait I double checked and it has access type UNKNOWN. Do we want to handle it as if it's ANY (and/or refactor to ANY?)

Comment (BrandonWeng) [reply app/graph.go]: Ah, I think they should be the same thing. If you feel that ANY is more clear then I can make a PR to change it

File: app/graph.go
Comment (BrandonWeng) [reply app/graph.go]: Gotcha, thanks for clarifying!

File: app/app.go
Comment (codchen) [inline app/app.go]: with this there will be two repetitive decoding for each tx (once here, once in `runMsg`). We don't need to optimize it in this PR but let's add a TODO here in case we forget

File: app/graph.go
Comment (codchen) [inline app/graph.go:189]: nit: it's not very easy to tell what `int` represents. Maybe we can define an explicit type `type DagNodeID int` for readability

**Review by BrandonWeng — APPROVED** [review]
lgtm, thanks for the change

**Pull request merged**
